### PR TITLE
Migrate exchange-related utility functions

### PR DESCRIPTION
**Pull request by bh2smith**
This is the first step in the cleanup task https://github.com/gnosis/dex-liquidity-provision/issues/184

This PR attempts to migrate any "exchange related utility functions" including

- `getUnitPrice(price, baseTokenDecmials, quoteTokenDecimals) -> PriceAsFraction`
- `getOutputAmountFromPrice(price, sellAmount, sellToken, buyToken) -> buyAmount`
- `getUnlimitedOrderAmounts(price, sellToken, buyToken) -> (sellAmount, buyAmount)`

I would suggest, that we generalize the function signature of `getOutputAmountFromPrice` to 

```js
getBuyAmountFromPrice(price, sellAmount, sellToken, buyToken)
```

~~I would also propose passing "Token" Types into these functions instead of just their decimals but am worried that this might make the fact that only the decimals are needed get lost.~~

### CODE REVIEW

File: src/util.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import { Fraction } from "./fraction";
+
+const max128 = new BN(2).pow(new BN(128)).subn(1);
+
+/**
+ * Modifies the price to work with ERC20 units
+ * @param {number} price amount of quote token in exchange for one base token
+ * @param {number} baseTokenDecimals number of decimals of the base token
+ * @param {number} quoteTokenDecimals number of decimals of the quote token
+ * @return {Fraction} fraction representing the amount of units of quote tokens in exchange for one unit of base token
```
Comment (nlordell) [inline src/util.ts]: nit: the types are no longer needed in the comments as they are specified in the function definition. Also the `tsdoc` standard expects a `-` between the parameter name and description.

```suggestion
 * @param price - Amount of quote token in exchange for one base token
 * @param baseTokenDecimals - Number of decimals of the base token
 * @param quoteTokenDecimals - Number of decimals of the quote token
 * @return Fraction representing the amount of units of quote tokens in exchange for one unit of base token
```

**Review by nlordell — APPROVED** [review]
Looks great! Just a minor nit about the [doc comment formatting](https://microsoft.github.io/tsdoc/).

File: src/util.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
+import BN from "bn.js";
+import { Fraction } from "./fraction";
+
+const max128 = new BN(2).pow(new BN(128)).subn(1);
```
Comment (fleupold) [inline src/util.ts]: ubernit: constant should be upper_cased?

File: src/util.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+ * @param {number} price amount of quote tokens in exchange for one base token
+ * @param {number} baseTokenDecimals number of decimals of the base token
+ * @param {number} quoteTokenDecimals number of decimals of the quote token
+ * @return {BN[2]} amounts of quote token and base token for an unlimited order at the input price
+ */
+export function getUnlimitedOrderAmounts(
+  price: number,
+  baseTokenDecimals: number,
+  quoteTokenDecimals: number,
+): [BN, BN] {
```
Comment (fleupold) [inline src/util.ts]: I assume you just copied this, but how about changing the return type to a named struct so that there is no ambiguity which value is for quote and which value is for base token?

**Review by fleupold — COMMENTED** [review]
Instead of having a util.ts file (which is likely going to become the dumping ground for any kind of helper function), I'd prefer if we can organize our helpers in modules that convey via their name what kind of helpers can be found (cf. onchain_reading, encoding, balance_reader).

Since this is mainly about price/amount conversion maybe it could be called `amounts` or something similar?

File: src/util.ts (hunk 3 above)
Comment (bh2smith) [reply src/util.ts]: I did not copy this, but If I understand correctly I could define a type called `SellAmount` and `BuyAmount` that are just BigNumbers so that people know exactly what is what.

**Comment by bh2smith** [comment]
> Since this is mainly about price/amount conversion maybe it could be called amounts or something similar?

I like the suggested term `amounts` here. I believe these functions were part of `price_utils.js` in the other project. Since we will not keep the "utils" part around and I don't thingk `prices.ts` is appropriate I will go with `amounts`.

**Comment by bh2smith** [comment]
> Looks great! Just a minor nit about the doc comment formatting.

In the doc comment you provided it appears we don't even need to provide the type (I assume this is because its given in the function signature).

File: src/util.ts (hunk 3 above)
Comment (nlordell) [reply src/util.ts]: You could just use an anonymous named struct:

```suggestion
): { base: BN, quote: BN } {
```

You could define a `SellAmount` and `BuyAmount` type alias we well:
```
type SellAmount = BN;
type BuyAmount = BN;
```

And while that would document the function a bit better, it still wouldn't prevent callers doing this:
```
const [quoteAmount, baseAmount] = estimateBuyAmount(...);
```

While, using the anonymous named struct you get:
```
const { quote, base } = estimateBuyAmount(...);
// exactly the same as:
const { base, quote } = estimateBuyAmount(...);
```

Also, you can still easily set the names in the destructuring with:
```
const { quote: quoteAmount, base: baseAmount } = estimateBuyAmount(...);
// exactly the same as:
const { base: baseAmount, quote: quoteAmount } = estimateBuyAmount(...);
```

And even ignore one of the values, without having to worry about their order:
```
const { quote } = estimateBuyAmount(...);
// with a named parameter
const { quote: quoteAmount } = estimateBuyAmount(...);
```

**Comment by nlordell** [comment]
>  (I assume this is because its given in the function signature).

Correct, it has no value in typescript. I suggest merging with master, as we added `eslint-plugin-tsdoc` to lint the documentation formatting.

**Comment by nlordell** [comment]
> Since we will not keep the "utils" part around and I don't thingk prices.ts is appropriate

I'm also not a big fan of `utils.*` files as they tend to turn into dumpig grounds for small methods that should probably be in their own modules.

I think, what we should do is create a module directory:
`src/utils/prices.ts`, that way other components can `import { getUnlimitedOrderAmounts } from "./utils/prices";`

**Comment by nlordell** [comment]
**edit**: Just to be clear, my comment was more about the directory structure, I'd be equally happy with `src/utils/amounts.ts`.

File: src/util.ts (hunk 3 above)
Comment (bh2smith) [reply src/util.ts]: Cool tips, I have gone with the following.

```js
type SellAmount = BN;
type BuyAmount = BN;

export function getUnlimitedOrderAmounts(
  price: number,
  sellTokenDecimals: number,
  buyTokenDecimals: number,
): { base: SellAmount; quote: BuyAmount } 
```

Does this make sense?

Comment (bh2smith) [reply src/util.ts]: I guess my next question would be with regard to declaring types in the same file and then trying to import them later for use in unit tests. Seems my linter is unhappy.

**Comment by bh2smith** [comment]
@fedgiac I would like to know if you had some example in mind that might be appropriate for unit testing some of these functions. I have been thinking in terms of their use cases for so long that I am, somewhat, blinded and can only think if using 18 or 6 decimal combinations. Also, noticed that more elaborate testing (like negative decimals) might imply we should impose some assertions/constraints (like decimals >= 0) in the functions themselves. This seems like a lot of work so I want to get an idea if this "worth the hassle".


In brief, some outstanding questions:

- Are we testing these utilities?
- If so, how intensively? 
- Should we impose further assertions on the input data (e.g. tokenDecimals be a non-negative integer)?
- If so, how extensively?

**Comment by fedgiac** [comment]
There already are some simple test cases for `getOutputAmountFromPrice` and `getUnlimitedOrderAmounts` in `test/price_utils.js`, so you could just take that. They require `toErc20Units` however, so maybe you need to rewrite the tests a bit. Also these functions are definitely not going to work with a negative number of decimals.

I'm not so sure about importing the `fetchTokenInfo*` functions, since retrieving token information isn't really something that has much to do with the exchange itself, also I don't much like how they work currently. Wasn't there the plan of having a repository containing utilities that are to be shared among different projects at Gnosis? They could fit there. Also `toErc20Units` and `fromErc20Units` are prime candidate for this use case.

File: src/util/amounts.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+ * @param buyTokenDecimals - Number of decimals of the buy token
+ * Returns Fraction representing the number of buy tokens in exchange for one unit of sell token
+ */
+export function getUnitPrice(
+  price: number,
+  sellTokenDecimals: number,
+  buyTokenDecimals: number,
+): Fraction {
+  assert(sellTokenDecimals > 0, "sell token decimals must be non-negative");
+  assert(buyTokenDecimals > 0, "buy token decimals must be non-negative");
```
Comment (fedgiac) [inline src/util/amounts.ts]: Zero decimals should be fine.
```suggestion
  assert(sellTokenDecimals >= 0, "sell token decimals must be non-negative");
  assert(buyTokenDecimals >= 0, "buy token decimals must be non-negative");
```

File: src/util/amounts.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  );
+}
+
+/**
+ * Computes the amount of output token units from their price and the amount of input token units
+ * Note that the price is expressed in terms of tokens, while the amounts are in terms of token units
+ * @param price - Amount of buy token in exchange for one sell token
+ * @param sellAmount - Amount of sell token units that are exchanged at price
+ * @param sellTokenDecimals - Number of decimals of the sell token
+ * @param sellTokenDecimals - Number of decimals of the buy token
```
Comment (fedgiac) [inline src/util/amounts.ts]:
```suggestion
 * @param buyTokenDecimals - Number of decimals of the buy token
```

File: src/util/amounts.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+ * Returns amount of output token units obtained
+ */
+export function getOutputAmountFromPrice(
+  price: number,
+  sellAmount: BN,
```
Comment (fedgiac) [inline src/util/amounts.ts]:
```suggestion
  sellAmount: SellAmount,
```

File: src/util/amounts.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  );
+  const buyTokenAmountFraction = unitPriceFraction.mul(
+    new Fraction(sellAmount, 1),
+  );
+  return buyTokenAmountFraction.toBN();
+}
+
+/**
+ * Computes the buy and sell token amounts required for an unlimited order in the exchange
+ * @param price - price of buy token in relative to one sell token
```
Comment (fedgiac) [inline src/util/amounts.ts]: Copied from the other two functions:
```suggestion
 * @param price - Amount of buy token in exchange for one sell token
```

File: src/util/amounts.ts (hunk 4 above)
Comment (bh2smith) [reply src/util/amounts.ts]: I'm still hesitating about whether I should even include these assertions. If we do, then I will have to test them.

Comment (nlordell) [reply src/util/amounts.ts]: I think assertions are nice to have in the code to demonstrate what are the developer's assumptions. I don't think you necessarily need to test them.

File: src/util/amounts.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+ */
+export function getBuyAmountFromPrice(
+  price: number,
+  sellAmount: SellAmount,
+  sellTokenDecimals: number,
+  buyTokenDecimals: number,
+): BuyAmount {
+  const unitPrice = getUnitPrice(price, sellTokenDecimals, buyTokenDecimals);
+  const buyAmount = unitPrice.mul(new Fraction(sellAmount, 1));
+  return buyAmount.toBN();
```
Comment (fedgiac) [inline src/util/amounts.ts:56]: Typescript question! The returned value is supposed to be of type `BuyAmount` but I'm fairly confident that `fraction.toBN()` is of type `BN`. Since there is no casting to `BuyAmount`, I expect TS to just treat `BuyAmount` as a synonym for `BN`. Are there practical implications of using a custom type or is it more akin to some sort of "type comment"?

File: test/models/util/amounts.spec.ts [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+import BN from "bn.js";
+import { assert } from "chai";
+import "mocha";
+
+const MAX128 = new BN(2).pow(new BN(128)).subn(1);
+const FLOAT_TOLERANCE = new BN(2).pow(new BN(52)); // This is float precision
+const ZERO_BN = new BN(0);
+const TEN_BN = new BN(10);
+const MAX8 = new BN(2).pow(new BN(8)).subn(1);
+const MAX256 = new BN(2).pow(new BN(256)).subn(1);
```
Comment (fedgiac) [inline test/models/util/amounts.spec.ts]: Very much nitpick: move `MAX128` between `MAX8` and `MAX256`.
```suggestion
const FLOAT_TOLERANCE = new BN(2).pow(new BN(52)); // This is float precision
const ZERO_BN = new BN(0);
const TEN_BN = new BN(10);
const MAX8 = new BN(2).pow(new BN(8)).subn(1);
const MAX128 = new BN(2).pow(new BN(128)).subn(1);
const MAX256 = new BN(2).pow(new BN(256)).subn(1);
```
(Interestingly, Github's diff algorithm doesn't minimize the number of lines changed.)

File: test/models/util/amounts.spec.ts [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+        expected: new Fraction(1, 10000000000),
+      },
+    ];
+    for (const {
+      price,
+      sellTokenDecmials,
+      buyTokenDecmials,
+      expected,
+    } of testCases) {
+      it(`evaluates as expected on input (${price}, ${buyTokenDecmials}, ${sellTokenDecmials})`, () => {
```
Comment (fedgiac) [inline test/models/util/amounts.spec.ts]: I like having the `it` inside the loop to see which test cases fail. :+1:

**Review by fedgiac — APPROVED** [review]
Looks good to me!
Some comments:
1. A known issue with `getBuyAmountFromPrice` is that it doesn't fail when the order cannot be placed at that price on the exchange. This should happen only in extreme cases (e.g., price is 1, token A has zero decimals, token B has 40 decimals, so 1 unit of token A is worth more than 2¹²⁸ units of token B), but still, trusting this function means that you'd be giving away your tokenB for zero tokenA. I don't know if we want to do something about this, since such an order would be simply impossible.
2. Including the full `toErc20Units` seems a bit overkill to me unless we plan to replace it with an import relatively soon. If not, much of its logic could be removed.
3. Typo: Decmials -> Decimals

File: src/util/amounts.ts (hunk 8 above)
Comment (bh2smith) [reply src/util/amounts.ts:56]: I am probably not the best to ask here on this, I would assume it is treated as a synonym.

**Comment by bh2smith** [comment]
Thanks for the feedback @fedgiac, 


> Including the full toErc20Units seems a bit overkill to me unless we plan to replace it with an import relatively soon. If not, much of its logic could be removed.

I tried to reduce this to just 

```js
return new BN(amount).mul(TEN_BN.pow(new BN(decimals)))
```
But some of our test cases failed, so I have stripped as much as possible without failing the existing tests.  This change also removed many of the constants that you had suggested moving around so that was easy.
